### ISSUE TITLE
q closes notmuch instead of returning to previous screen

### ISSUE THREAD

**Issue by TheKashe** [cmiles74/spacemacs-notmuch-layer#3]
pressing "q" should return to the previous screen but instead it closes notmuch completely. The workaround is adding the below line to config
`
 (push "\\*notmuch-.+\\*" spacemacs-useful-buffers-regexp)
`
This should be added to the layer out of the box.

**Comment by cmiles74** [comment]
My understanding is that when you are in any notmuch related buffer (the most common case being that your reading an email message), when you press the `q` key that buffer should close (that is, you don't want a bunch of email message buffers lying around for emails you don't care much about). But you are right that this works in _any_ notmuch related buffer, including a search or even the "hello" buffer.

I think this behavior works well when reading email messages, I haven't found it much of a bother in the other modes but my usage really revolves around reading messages. Are you thinking that `q` should only close message buffers and, in other modes, it should just push notmuch to the back? On my machine notmuch opens to my inbox search pretty quickly, that said I only have ~900 messages in my inbox.

Thank you!

**Comment by TheKashe** [comment]
"q" should go back to the previous buffer when used anywhere within notmuch. With spacemacs that behaviour doesn't seem to work. 

I'm completely new to spacemacs and notmuch so I'm trying to put he pieces of the puzzle together, but apparently spacemacs has something called [useful buffers](https://github.com/syl20bnr/spacemacs/issues/6683). The problem is ,notmuch buffer names start with "*", which spacemacs doesn't treat as useful, hence closing the current buffer takes you to the next useful buffer, which is non-notmuch.

Fix from the first post makes notmuch buffers "useful", which makes "q" behave as it should. Another option is `(setq spacemacs-useless-buffers-regexp '("^ "))`.

**BUT**, it breaks again if you switch layers :-(

Scenarion:
- set the above option in .spacemacs
- restart
- open notmuch
- open one of the saved searches
- open one of the messages
- press "q" - you are now back to *notmuch-saved-search...
- press "q" - you are now at *notmuch-hello*
- repeat as desired, "q" works properly
- open one of the saved searches
- change layer, e.g. SPC-l-2
- change layer back, e.g. SPC-l-1
- you now still see the *notmuch-saved-search buffr
- press "q"
- it will go to one of the non-notmuch layers, closing *notmuch-hello* :-(
- and it appears to stay broken until restart

Another issue is closing the help screen. Open it with ?. "q" should take you back, but it doesn't. 

[Reference to a similar problem ](https://github.com/syl20bnr/spacemacs/issues/6787)(but not mentioning layers)

**Comment by TheKashe** [comment]
Another way to check this:
- add `(setq spacemacs-useless-buffers-regexp '("^ "))` to config
- open *scratch*, kill all buffers so you start clean
- open notmuch
- go to search, read a message
- in this view, hit SPC-b-. to open buffer transient state
- move with n and p, you'll notice notmuch buffers are there
- open another layer SPC-l-1
- go back SPC-l-1
- hit SPC-b-. -> notmuch buffers are gone!

It appears that switching layers resets `spacemacs-useless-buffers-regexp`, but printing it out still shows the same for me, so i'm really puzzled as to what happens here.

**Comment by TheKashe** [comment]
I think it's a bug in spacemacs itself, filed a bug https://github.com/syl20bnr/spacemacs/issues/10991.

While this is opened, the only possible workaround would be to rename notmuch layers to not include stars. But that's hardcoded in notmuch, so probably not possible :-(

Edit: also filed in persp-mode. There is `*` hardcoded in persp-mode, and there are `*` hardcoded in notmuch. The more I try to workaround, the worse it gets. The only feasible workaround is running notmuch in it's own frame :-(

**Comment by jernejkase** [comment]
Adding the below hook makes useful buffers behave correctly after switch again, but not until all buffers are initialised.

I think it comes down to perspective switching - probably buffers with asterisk are not saved before switch

` (add-hook 'persp-switch-hook(lambda ()(setq "^" spacemacs-useful-buffers-regexp)))`

**Comment by TheKashe** [comment]
Ahhh.... I think we need [custom saving/loading functions](https://github.com/Bad-ptr/persp-mode.el#custom-saveload-buffer-function-example)!

**Comment by TheKashe** [comment]
I think it's a bug in persp-mode. Notmuch doesn't seem to preserve order of buffers

**Comment by TheKashe** [comment]
It appears the only way to reorder buffers in emacs is to redisplay them. So in order to have the notmuch layers in the right order after swithc, we need to do sth like the concept below (it assumes notmuch perspective is named "notmuch" )

```
(defun jk-switch (buff)
    (switch-to-buffer buff)
    (sit-for 0) ;;redisplay)
  (defun jk-persp-after-switch (frame)
    (when  (string= "notmuch" (safe-persp-name (get-current-persp)))
      (mapcar #'jk-switch (reverse (safe-persp-buffers (frame-parameter (selected-frame) 'persp))))))
(add-hook 'persp-activated-functions 'jk-persp-after-switch)
```
There is flickering when perspective is redisplayed, perhaps we could open a small window and open the buffers there.

Now another issue is that apparently notmuch is not meant to work in parallel windows. If I open one saved search in one window  and the other in another, the other is locked until the first one is released.

**Comment by TheKashe** [comment]
[Eyebrowse](https://github.com/wasamasa/eyebrowse) is use used for managing workspaces, which is entirely different beast. Biggest problem is that it doesn't keep buffer list the way persp-mode does. In essence, all buffers live in persp-mode, and eyebrowse just brings up the visible windows and their buffers, so there is no buffer list or anything like that.

I think the only feasible way to solve this on the spacemacs level would be to also use persp-mode to manage windows (actually I wonder why that is not the case?).

But perhaps we should look in completely different direction and maintain buffer stack within notmuch and return to the previous buffer with q, rather than just kill the current buffer.

**Comment by cmiles74** [comment]
It sounds like this is a problem with the Notmuch mode in Emacs and not necessarily with this layer. I'm going to close this for now, but please let me know if you finally get this one resolved to your satisfaction.

Thank you! `:-)`
